### PR TITLE
Robust bundle source rewrite

### LINKED ISSUE #1249: bundle-source: trailing line comment eats all exports
**Describe the bug**

@Chris-Hibbert and I tracked down a problem with the ZCF code in which the primary symptom was a dynamic `createVat` which failed because `buildRootObject` was not an export of the vat definition file. This file happened to end with a line comment:

```js
export function buildRootObject(vatPowers) {
 ...
}
// last line of file
```

The generated bundle takes that source and wraps it in some boilerplate. The prologue provides a CJS-style `module` and `module.exports` object. The epilogue contains `exports.buildRootObject = buildRootObject;` for each name exported by the file. The resulting string is built like `prologue + definition + epilogue`.

The fun part is that all the whitespace at the end of the definition file seems to have gone missing. Because the missing newline no longer terminates the `//` comment at the end of the definition, the comment now includes all of the `exports` assignments. As a result, the namespace object is empty, except for a `__esModule` property that was assigned in the prologue.

Multiple newlines go missing too. Adding a single semicolon, or a `/* comment */` at the end resolves the issue.

**To Reproduce**

The shortest path to reproduce is to edit `packages/SwingSet/test/metering/metered-dynamic-vat.js` to add `// comment` to the end of the file. Then run `node -r esm test/metering/test-dynamic-vat.js`.

The more complete path is:

Create a vat definition file like:

```js
export function buildRootObject() {
  return harden({ run() { return 0; } });
}
// comment

```

Turn it into a bundle with `bundleSource`. Get the bundle into a static vat somehow. Feed the bundle to `E(vatAdmin).createVat(bundle)`.

The `createVat` will fail, with an error message that mentions `vatNS does not have a property named 'buildRootObject'`.

**Workarounds**

The most expedient workaround is to not put a `//` line comment at the end of vat definition files. Use a block comment (`/* like this */`), or add a stray trailing semicolon if eslint will allow it.

cc @michaelfig who might know where this is happening in bundle-source, and @Chris-Hibbert who hit the bug originally

### PR DESCRIPTION
**Pull request by michaelfig**
Fixes #1249

This PR reimplements the following bundle-source features in a more robust way:

- Prevent many false positives (SES failing because of import expression detected, or HTML comment) by rewriting JS comments in the bundled source.
- Unmap the bundled source chunks so that expressions are present on the line numbers corresponding to the original sources.  We use only the start position of each AST node to decide the line number on which they should appear.

We do this by using the Babel parser and generator we already require for the tildot plugin.

### CODE REVIEW

**Review by warner — APPROVED** [review]
looks good, I'm kinda curious about the need for the vat-dave change, but feel free to land.

**Comment by michaelfig** [comment]
I'll probably withdraw this unless I can get more predictable behavior from babel.

**Comment by erights** [comment]
> I'll probably withdraw this unless I can get more predictable behavior from babel.

I saw this but have not looked at the context. What is the less predictable behavior from babel currently?

**Comment by michaelfig** [comment]
> > I'll probably withdraw this unless I can get more predictable behavior from babel.
> 
> I saw this but have not looked at the context. What is the less predictable behavior from babel currently?

Babel generator has a `retainLines: true` option which is supposed to honour the AST's line numbers when outputting generated code.  I've found this to be hit-and-miss, at least with the ASTs I've been feeding it.

File: packages/zoe/test/swingsetTests/zoe/vat-dave.js [hunk 1]
```diff
@@ -85,7 +85,7 @@ const build = async (E, log, zoe, issuers, payments, installations, timer) => {
       const installationCode = await E(zoe).getInstallation(installationHandle);
       // pick some arbitrary code points as a signature.
       assert(
-        installationCode.includes('asset: give.Asset,'),
+        installationCode.match(/asset: give\.(\\n|\s)*Asset,/s),
```
[the remark this answers is not in the document]
Comment (michaelfig) [reply packages/zoe/test/swingsetTests/zoe/vat-dave.js]: This is indicative of a bigger problem I'm having with the rewrite... it doesn't seem to want to preserve line numbers even though I'm putting the AST in a shape that looks like it should.

File: packages/zoe/test/swingsetTests/zoe/test-zoe.js [hunk 2]
```diff
@@ -30,7 +30,7 @@ const generateBundlesP = Promise.all(
 );
 
 async function main(withSES, basedir, argv) {
-  const dir = path.resolve('test/swingsetTests', basedir);
+  const dir = path.resolve(`${__dirname}/..`, basedir);
```
[the remark this answers is not in the document]
Comment (michaelfig) [reply packages/zoe/test/swingsetTests/zoe/test-zoe.js]: Done in #1191

File: packages/zoe/test/swingsetTests/zoe/vat-dave.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (michaelfig) [reply packages/zoe/test/swingsetTests/zoe/vat-dave.js]: Victory!  The rewrite is working.  I'm now getting expected output.

**Comment by michaelfig** [comment]
> Babel generator has a `retainLines: true` option which is supposed to honour the AST's line numbers when outputting generated code. I've found this to be hit-and-miss, at least with the ASTs I've been feeding it.

It turned out to be undesired aliasing of "location" objects in my transform.  Changing this to take copies of them made it work as expected.

**Review by warner — APPROVED** [review]
I don't know how babel works, so I can't comment on the implementation. But the tests look good, and I'm glad to see one that checks the trailing-comment bug.

I *think* the HTML comment and dynamic-import-in-comment edits are harmless, especially if I'm reading that right and those rewrites are only applied inside the parsed JS comment spans. I'd like to get @erights opinion on making those changes, though, because I know he was concerned about replacing surprising behavior inside the JS engine with surprising behavior inside `bundleSource`. So consider my review conditional upon his approval of that approach.

**Comment by michaelfig** [comment]
> I _think_ the HTML comment and dynamic-import-in-comment edits are harmless, especially if I'm reading that right and those rewrites are only applied inside the parsed JS comment spans.

That's correct.

File: packages/bundle-source/src/index.js [hunk 3]
```diff
@@ -73,51 +74,79 @@ export default async function bundleSource(
     if (isEntry) {
       entrypoint = fileName;
     }
-    let unmappedCode = code;
+
+    // Parse the rolled-up chunk with Babel.
+    const ast = (babelParser.parse || babelParser)(code);
```
Comment (kriskowal) [inline packages/bundle-source/src/index.js:80]: What causes the type of `babelParser` to vary? Maybe leave a comment.

File: packages/bundle-source/src/index.js [hunk 4]
```diff
@@ -73,51 +74,79 @@ export default async function bundleSource(
...
-      .replace(IMPORT_RE, '$1notreally')
-      .replace(HTML_COMMENT_RE, '<->');
-    // console.log(`<<<<<< ${fileName}\n${defangedCode}\n>>>>>>>>`);
-    sourceBundle[fileName] = defangedCode;
+    const rewriteComment = node => {
+      node.type = 'CommentBlock';
+      node.value = node.value
+        .replace(/^\s*/gm, '') // Strip extraneous comment whitespace.
+        .replace(HTML_COMMENT_START_RE, '<!X-')
+        .replace(HTML_COMMENT_END_RE, '-X>')
```
Comment (kriskowal) [inline packages/bundle-source/src/index.js:123]: Could use a comment like “Here we replace markers that resemble HTML comments within JS comments like `// <!--` with `// <!X--` so they can pass through the SES bans on risky source patterns.

**Review by kriskowal — APPROVED** [review]

File: packages/bundle-source/src/index.js (hunk 4 above)
Comment (michaelfig) [reply packages/bundle-source/src/index.js:123]: I can't put those strings in comments or they'll fail under SES. 😝 

I can (and have) described it with prose, though.

File: packages/bundle-source/src/index.js (hunk 3 above)
Comment (michaelfig) [reply packages/bundle-source/src/index.js:80]: Done.  Module import systems.

**Review by erights — APPROVED** [review]
LGTM for the part I was asked to look at. I have not reviewed the rest.

**Pull request merged**